src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1137 @@
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+        self.check_pin_hash_enc(rng, persistent_store, &aes_dec_key, pin_hash_enc)?;
+
+        // Assuming PIN_TOKEN_LENGTH % block_size == 0 here.
+        let iv = [0u8; 16];
+        let mut blocks = [[0u8; 16]; PIN_TOKEN_LENGTH / 16];
+        for (i, item) in blocks.iter_mut().take(PIN_TOKEN_LENGTH / 16).enumerate() {
+            item.copy_from_slice(&self.pin_uv_auth_token[i * 16..(i + 1) * 16]);
+        }
+        cbc_encrypt(&aes_enc_key, iv, &mut blocks);
+        let mut pin_token = vec![];
+        for item in blocks.iter().take(PIN_TOKEN_LENGTH / 16) {
+            pin_token.extend(item);
+        }
+
+        #[cfg(feature = "with_ctap2_1")]
+        {
+            self.permissions = 0x03;
+            self.permissions_rp_id = None;
+        }
+
+        Ok(AuthenticatorClientPinResponse {
+            key_agreement: None,
+            pin_token: Some(pin_token),
+            retries: None,
+        })
+    }
+
+    #[cfg(feature = "with_ctap2_1")]
+    fn process_get_pin_uv_auth_token_using_uv_with_permissions(
+        &self,
+        _: CoseKey,
+    ) -> Result<AuthenticatorClientPinResponse, Ctap2StatusCode> {
+        // User verifications is only supported through PIN currently.
+        #[cfg(not(feature = "with_ctap2_1"))]
+        {
+            Err(Ctap2StatusCode::CTAP1_ERR_INVALID_COMMAND)
+        }
+        #[cfg(feature = "with_ctap2_1")]
+        {
+            Err(Ctap2StatusCode::CTAP2_ERR_INVALID_SUBCOMMAND)
+        }
+    }
+
+    #[cfg(feature = "with_ctap2_1")]
+    fn process_get_uv_retries(&self) -> Result<AuthenticatorClientPinResponse, Ctap2StatusCode> {
+        // User verifications is only supported through PIN currently.
+        #[cfg(not(feature = "with_ctap2_1"))]
+        {
+            Err(Ctap2StatusCode::CTAP1_ERR_INVALID_COMMAND)
+        }
+        #[cfg(feature = "with_ctap2_1")]
+        {
+            Err(Ctap2StatusCode::CTAP2_ERR_INVALID_SUBCOMMAND)
+        }
+    }
+
+    #[cfg(feature = "with_ctap2_1")]
+    fn process_set_min_pin_length(
+        &mut self,
+        persistent_store: &mut PersistentStore,
+        min_pin_length: u8,
+        min_pin_length_rp_ids: Option<Vec<String>>,
+        pin_auth: Option<Vec<u8>>,
+    ) -> Result<(), Ctap2StatusCode> {
+        if min_pin_length_rp_ids.is_some() {
+            return Err(Ctap2StatusCode::CTAP2_ERR_UNSUPPORTED_EXTENSION);
+        }
+        if persistent_store.pin_hash().is_some() {
+            match pin_auth {
+                Some(pin_auth) => {
+                    if self.consecutive_pin_mismatches >= 3 {
+                        return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_BLOCKED);
+                    }
+                    let mut message = vec![0xFF; 32];
+                    message.extend(&[0x06, 0x08]);
+                    message.extend(&[min_pin_length as u8, 0x00, 0x00, 0x00]);
```

Comment:
Can you document what these magic values mean (maybe with some reference to the specifications), given that they don't seem to use CBOR encoding but some custom format instead?

Reply:
This is tricky, since the currently public CTAP 2.1 revision
https://fidoalliance.org/specs/fido2/fido-client-to-authenticator-protocol-v2.1-rd-20191217.html
is from December 2019, and `SetMinPinLength` is in the internal-only version.

I added a TODO to improve the documentation when possible.

Reply:
No, this is a cryptographic trick to make `pin_auth` codes not reusable between commands.

The longer explanation: CTAP 2.1 introduces the concept of permissions for tokens. So when you enter your PIN, you still get a token, but it only valid for a specific use case. To enforce that, reply attacks are supposed to be prevented by making the hashed message start with a unique prefix per command.

Since this cryptographic feature did not exist before, they start with `[0xFF; 32]` to distinguish form old commands. Then, they add one byte for the command (0x06 == clientPin) and one byte for the subcommand (0x08 == minPinLength).

In summary I think this is a backwards compatible attempt to defend against an attack that wasn't considered before. This level of explanation is not even given in the command description, so I assumed this source code isn't the place to discuss cryptographic protocol details as well?

Reply:
Hum "cryptographic trick" triggers my reviewer scrutiny ^^ Though after all I don't see much cryptography in your explanation.

So the format is `[0xFF; 32]` + a regular encoded command/subcommand, right? Can we somehow use the command/subcommand serializer to create these bytes then (and using constants/enums that explicitly mention "clientPin" and "minPinLength" rather than the unexaplained `0x06` and `0x08`)?

Reply:
I think quoting this sentence from the specs will be the clearest option then! I didn't realize it was that bad in the specs...